doc/design/proposal-noderc.md

```diff
@@ -0,0 +1,164 @@
+# On-disk Node.js runtime configuration (noderc)
+
+The motivation is to provide a way for users to specify runtime configuration for a Node.js application in an on-disk file that will be discovered automatically by Node.js.
+
+For background and early discussions about the format, etc., see https://github.com/nodejs/node/issues/53787
+
+## Overview
+
+### `"noderc"` in package.json
+
+An on-disk runtime configuration should be first specified in a relevant `package.json` file using the `"noderc"` field. A `package.json` file considered relevant when it's placed in any directory from the root directory to the base directory. The base directory is either the directory where the Node.js application entry point file is (if Node.js is launched to execute a file), or the current working directory (e.g. if Node.js is launched as a REPL). An example `package.json` file looks like this:
+
+```json
+{
+  "name": "my-project",
+  "schema": "1.0.0",
+  "noderc": "./.noderc.json"
+}
+```
+
+In the initial iteration, the runtime configuration file must be in JSON format, and has an extension that ends with `.json`. We are open to support more formats identified by other extensions in the future but the details will remain to be discussed and it won't be implemented in the initial iteration.
+
+The idiomatic way to specify this configuration would be using a file named `.noderc.json` in the same directory as the `package.json`, which tends to be in the project root directory.
+
+
+### Basic of `noderc` in JSON
+
+```json
+{
+  "schema": 0,
+  "import": [ "amaro/register", "./monitor.ts" ]
+}
+```
+
+- The `schema` field is only meant for breaking changes to the schema, so it's a single number.
+  - At schema 0, there is no stability guarantee about the schema
+  - When we iterate on the schema to a point to consider it stable, the schema will be set to 1
+  - If the noderc is using a schema that's not the latest supported by the running Node.js version (e.g. it uses 2 but the latest schema supported by the Node.js version is 3), Node.js converts the older schema to the newer schema in the underlying implementation.
+  - If the noderc is using a schema that the current Node.js version doesn't support (e.g. it's schema 3 while the Node.js version only supports up to 2), an error is thrown.
```

Comment:
Id also specify behavior of extending configuration files with different schema version

Reply:
I think maybe it's okay to do the conversion on a per-file basis? Though initially, to just keep things simple, we can just throw an error for mismatching schema extensions.

Reply:
Put it in the reusing section as "throw initially, in future iterations, maybe allow conversions and merging - depending on how stable the schemas turn out to be"